INSERT...SELECT with re-partitioning

pykello opened the pull request:
DESCRIPTION: INSERT...SELECT with re-partitioning

We currently have two methods of executing an INSERT...SELECT: 
- in parallel between co-located shards, used when the distribution column value is preserved between source and target table
- via the coordinator using COPY, used in all other cases

The first method scales horizontally, but the second method can get bottlenecked on single core performance on the coordinator, and may cause the coordinator to run out of disk space.

This PR adds a third method:
- distributed repartitioning of the select result, used in cases where the SELECT is a distributed query that does not require a merge on the coordinator

The INSERT..SELECT runs in 3 phases:
1. Run SELECT tasks wrapped in `worker_partition_query_result` to partition query results into fragments which correspond to the hash ranges of the shards in the destination table
2. Fetch fragments to the corresponding destination shard using `fetch_intermediate_results()` using one call per node pair
3. Run `INSERT INTO .. SELECT .. FROM read_intermediate_results(..)` into the shards of the destination table, where read_intermediate_result reads the concatenation of fragments that correspond to the hash range of the shard

In each phase, the tasks are executed by the adaptive executor. Empty tasks are pruned away during execution.

Earlier prototype was https://github.com/citusdata/citus/pull/3267

Closes #2524

marcocitus commented on src/backend/distributed/executor/insert_select_executor.c:
```diff
@@ -144,8 +164,102 @@ CoordinatorInsertSelectExecScanInternal(CustomScanState *node)
+		 * We don't support insert/select with repartitioning if:
+		 *  (*) Query is not distributed or requires merge on coordinator, or
+		 *  (*) Has returning, or
+		 *  (*) Has ON CONFLICT
```
What's the issue with ON CONFLICT?

pykello replied:
No issues, `multi_insert_select_conflict` was failing because I hadn't added casts for SELECT target list, so rows weren't routed correctly and it was failing. For example, if target relation was partitioned by hash(int), we were partitioning the SELECT correctly by hash(numeric).

Fixed that and enabled ON CONFLICT.

marcocitus commented on src/backend/distributed/executor/insert_select_executor.c:
```diff
@@ -595,3 +680,219 @@ ExecutingInsertSelect(void)
+
+		/* sort result ids for consistent test output */
+		List *sortedResultIds = SortList(resultIdList, pg_qsort_strcmp);
+
```
Maybe skip the remainder if sortedResultIds is empty.

pykello replied:
done

pykello commented:
@marcocitus I am adding more tests and fixes as I do more testing, but feel free to review what is there so far. Thanks.

marcocitus commented on src/backend/distributed/executor/insert_select_executor.c:
```diff
@@ -595,3 +681,362 @@ ExecutingInsertSelect(void)
+static bool
+IsRedistributablePlan(Plan *selectPlan, bool hasReturning)
+{
+	if (hasReturning)
```
What's the issue with returning?

pykello replied:
enabled at and added a test.

marcocitus commented on src/backend/distributed/executor/insert_select_executor.c:
```diff
@@ -144,8 +179,101 @@ CoordinatorInsertSelectExecScanInternal(CustomScanState *node)
+			TransactionProperties xactProperties = {
+				.errorOnAnyFailure = true,
+				.useRemoteTransactionBlocks = TRANSACTION_BLOCKS_REQUIRED,
+				.requires2PC = false
```
Should this be true?

pykello replied:
Changed to using `ExecuteTaskListIntoTupleStore()` which decides on the correct value without explicitly assigning it.

marcocitus commented on src/backend/distributed/executor/partitioned_intermediate_results.c:
```diff
@@ -179,6 +180,23 @@ worker_partition_query_result(PG_FUNCTION_ARGS)
+		/*
+		 * Caller of this function might request to reduce number of output
+		 * columns, because they don't want to get the target entries added
+		 * in ExpandWorkerTargetEntry() in the output.
```
Are there any alternative to solving it here?

pykello replied:
> Are there any alternatives to solving it in this function?

The best way would be refactor planner to remove extra columns if there is no coordinator work to do. I'll work on that today.

Other alternative is wrapping the task query and selecting only columns we need, etc. might be a little brittle as I need to make assumptions about how the planner assigns names, etc.

pykello replied:
Wrapped task query to select only columns we are interested in. I wasn't aware of `job->jobQuery` which can be used to get list of columns assigned by citus planner.

marcocitus commented:
Would be nice if explain output reflected re-partitioning instead of "via coordinator":
```

postgres=# explain INSERT INTO test (y, x) SELECT * FROM test;
┌─────────────────────────────────────────────────────────────────────────────────────────┐
│                                       QUERY PLAN                                        │
├─────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus INSERT ... SELECT via coordinator)  (cost=0.00..0.00 rows=0 width=0) │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)                    │
│         Task Count: 32                                                                  │
│         Tasks Shown: One of 32                                                          │
│         ->  Task                                                                        │
│               Node: host=localhost port=9700 dbname=postgres                            │
│               ->  Seq Scan on test_102008 test  (cost=0.00..32.60 rows=2260 width=8)    │
└─────────────────────────────────────────────────────────────────────────────────────────┘
(7 rows)
```

Some tests that might be worth adding:
- prepared statements
- select is a delegated function call
- from worker node in mx

pykello commented:
> Would be nice if explain output reflected re-partitioning instead of "via coordinator":

Changing the title would require quite bit of refactoring, so I changed the title and added a text property to specify which INSERT SELECT method is being used.

pykello commented:
@marcocitus today:
1. Updated the EXPLAIN output
2. Added MX, function delegation, failure, and prepared statement tests

plan for before merge:
1. Couple of more range partitioned tests
2. error and rollback behaviour tests
3. Some replicated target tests
4. Any meaningful isolation tests

marcocitus commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -150,6 +154,17 @@ CoordinatorInsertSelectExplainScan(CustomScanState *node, List *ancestors,
+	}
+	else
+	{
+		ExplainPropertyText("INSERT/SELECT method", "route via coordinator", es);
```
Maybe "pull to coordinator". 

"route" is probably more correct, but somehow seems a bit less clear, mostly because it makes me think of router queries.

pykello replied:
done

pykello commented:
@marcocitus I don't have much to add. Today I added the tests I intended, changed "via coordinator" to "pull to coordinator", and created a copy of select query before planning in EXPLAIN, since planner changes the query and we need to plan it again later.

marcocitus commented:
I think other than the above issue and the GUC, this looks good to me.

pykello commented:
@marcocitus addressed both points.

pykello commented:
@marcocitus today:
1. Added a GUC in https://github.com/citusdata/citus/pull/3376/commits/7088f85a6c47246a9f77098e8625a851ce100899
2. To fix the issue you raised with ambiguous column names, relabeled select columns to match insert target list in https://github.com/citusdata/citus/pull/3376/commits/8239798fada66d9890a9ee0e1cbc67f21461741d
3. Added jobid to result id prefix, since INSERT/SELECT can be recursively planned: https://github.com/citusdata/citus/pull/3376/commits/d5702835f2031a501dfeffd113953886bd3177f1 (it seemed to work without this change too in some cases I tried, but it was much harder to reason about why that'd work correctly)

marcocitus commented:
Ran into a general (pre-existing) issue with our handling of RETURNING in INSERT..SELECT: https://github.com/citusdata/citus/issues/3395

We seem to reject such queries in update/delete/insert, but not in insert..select.

marcocitus commented:
Issue with re-partition joins:
```sql
postgres=# INSERT INTO test (y, x) SELECT a.x, b.y FROM test a JOIN test b USING (y);
ERROR:  relation "pg_merge_job_2133675999233.task_000033" does not exist
CONTEXT:  while executing command on localhost:9701
```

pykello commented:
> Issue with re-partition joins:
> 
> ```sql
> postgres=# INSERT INTO test (y, x) SELECT a.x, b.y FROM test a JOIN test b USING (y);
> ERROR:  relation "pg_merge_job_2133675999233.task_000033" does not exist
> CONTEXT:  while executing command on localhost:9701
> ```

Disabled repartitioned insert/select for them for now: https://github.com/citusdata/citus/pull/3376/commits/80aa39b3869f40ee9fe1cd6f68259326818ceffe
